### PR TITLE
Pipeline build steps

### LINKED ISSUE #4956: Deprecate git hooks
**What is the goal / desired outcome?** Mark git hooks as deprecated, and remove in pachyderm 2.0. This shouldn't be pursued [until build steps are complete.](https://github.com/pachyderm/pachyderm/issues/4958)

**If there is a way to accomplish this today via workaround, what does that require?** We'll release two alternatives:
1. [A build step for pachyderm pipelines](https://github.com/pachyderm/pachyderm/issues/4958)
2. [A demonstration of making a git hook equivalent using a pachyderm spout+service pipeline](https://github.com/pachyderm/pachyderm/issues/4957)

### LINKED ISSUE #4958: Add a build step for pachyderm pipelines
**What is the goal / desired outcome?** We plan to add basic support for build steps in pachyderm pipelines. This will draw from [`create_python_pipeline`](https://pachyderm.github.io/python-pachyderm/python_pachyderm.m.html#python_pachyderm.create_python_pipeline), but be more language-agnostic, built into `pachctl`, and a little cleaner (by avoiding a separate source repo and build pipeline.)

**If there is a way to accomplish this today via workaround, what does that require?**

1. Users can bake docker images and go through the normal pipeline update process.
2. Git hooks, though [they'll be removed shortly.](https://github.com/pachyderm/pachyderm/issues/4956)

### PR DESCRIPTION
**Pull request by ysimonson**
## Motivation

We want to make it easier for people to quickly develop and iterate on Pachyderm pipelines. At the moment, users need to:

1. Make changes to their code
2. Build a docker image
3. Push the docker image
4. Update the pipeline

We do offer some conveniences around this workflow (`pachctl update pipeline --build`), but the process is still too cumbersome. Additionally, the current setup requires docker, which doesn't work in all environments (including the pachyderm IDE) and is sometimes unfamiliar for data scientists.

## Overview

Adds the notion of a "build step" to pipelines, which acts as a cleaner, language-agnostic, `pachctl`-level variant of python-pachyderm's [`create_python_pipeline`](https://pachyderm.github.io/python-pachyderm/python_pachyderm.m.html#python_pachyderm.create_python_pipeline). When a build step-enabled pipeline is created or updated:

1. The pipeline's code is uploaded to a branch.
2. A pipeline is created that pulls dependencies and compiles code from the source.
3. The specified pipeline is created or updated, with its specified inputs as well as the newly created source and build assets.

Creating a DAG like this:

```
.---------------------------------------------------------.
| repo "<pipeline name>_build"                            |
|                                                         |
| .-----------------.      .----------------------------. |
| | branch "source" | ---▶ | pipeline on branch "build" | |
| `-----------------'      `----------------------------' |
`----------|------------------------|---------------------'
           |                        |
           | /pfs/source            | /pfs/build
           ▼                        |
  .----------------------------.    |
  | pipeline "<pipeline name>" | ◀--'
  `----------------------------'
           ▲
           | /pfs/<input name(s)>
           |
      .------------.
      | <input(s)> |
      '------------'
```

## Pipeline spec

This feature adds a `build` object to the pipeline spec's `transform` object. It has the following fields:

* `path`: An optional string specifying where the source code is relative to the pipeline spec path (or the cwd if the pipeline is fed into `pachctl` via stdin.)
* `language`: A string specifying what language builder to use (see below.) Only works with official builders.
* `image`: A string specifying what builder image to use, if a non-official builder is desired.

Either `language` or `image` must be specified. If `transform.image` is not specified, it's set to be the same as the builder image, which should work in most cases. Similarly, if `transform.cmd` is not specified, it will default to `sh /pfs/build/run.sh`, which should be provided via the build pipeline.

This example will create a pipeline that builds go code located in the `source` directory:

```json
{
  "pipeline": {
    "name": "map"
  },
  "description": "A pipeline that tokenizes scraped pages and appends counts of words to corresponding files.",
  "transform": {
    "build": {
      "language": "go",
      "path": "./source"
    }
  },
  "input": {
    "pfs": {
      "repo": "scraper",
      "glob": "/*"
    }
  }
}
```

This example uses a custom builder for julia code that is located in the same directory as the pipeline spec:

```json
{
  "pipeline": {
    "name": "map"
  },
  "description": "A pipeline that tokenizes scraped pages and appends counts of words to corresponding files.",
  "transform": {
    "build": {
      "image": "ysimonson/pachyderm-julia-builder:latest",
    }
  },
  "input": {
    "pfs": {
      "repo": "scraper",
      "glob": "/*"
    }
  }
}
```

When a user creates or updates a pipeline with a build spec, the build process outlined above is automatically triggered.

## Source code

Requirements on how source code is structured is language-specific. When using the official builders:

* For python:
  * There must be a `main.py` in the source root
  * There can be an optional `requirements.txt` specifying pip dependencies
* For go:
  * There must be one or more go source files who's package is `main` in the source root

The only file that applies to all builders is `.pachignore`. It's an optional file in the source root containing [ohmyglob](https://github.com/pachyderm/ohmyglob) entries specifying which files should not be uploaded to PFS.

## Builders

Official builders are available for `python` and `go`. These can be used by setting `transform.build.language` in a pipeline spec.

For other languages, or customizations to the official builders, users can author their own builders. Builders are somewhat similar to [buildpacks](https://docs.cloudfoundry.org/buildpacks/) in design, and follow a convention over configuration approach. A builder needs 3 things:

* A `Dockerfile` to bake the image.
* A `build.sh` in the image workdir, which acts as the entry-point for the build pipeline.
* A `run.sh`, injected into `/pfs/out` via `build.sh`. This will act as the entry-point for the executing pipeline. By convention, `run.sh` should take an arbitrary number of arguments and forward them to whatever executes the actual user code.

A custom builder can be used by setting `transform.build.image` in a pipeline spec. The official builders can be used for reference; they're located in `etc/pipeline-build`.

## Other changes

* Deprecated git hooks
* Added some (minimal) client-side validation of pipeline specs to prevent panics

Closes #4958 
Closes #4956

### CODE REVIEW

File: src/client/pps/pps.proto [hunk 1]
```diff
@@ -34,6 +34,13 @@ message Transform {
   string user = 10;
   string working_dir = 11;
   string dockerfile = 12;
+  BuildSpec build = 15;
+}
+
+message BuildSpec {
+  string path = 1;
+  string language = 2;
```
Comment (jdoliner) [inline src/client/pps/pps.proto:42]: I'm guessing that language is meant to add some extra ergonomics to this feature by giving commonly used languaes prefab build containers. What's actually in those images, do they include Pachyderm clients in addition to the standard language containers?

File: src/server/pps/cmds/cmds.go [hunk 2]
```diff
@@ -1071,191 +1075,391 @@ you can increase the amount of memory used for the bloom filters with the
...
+		}
+	}
+
+	// insert the source code
+	commit, err := pc.StartCommit(buildPipelineName, "source")
+	if err != nil {
+		return errors.Wrapf(err, "failed to start commit for source code in build step-enabled pipeline")
+	}
+	if update {
+		if err = pc.DeleteFile(buildPipelineName, commit.ID, "/"); err != nil {
```
Comment (jdoliner) [inline src/server/pps/cmds/cmds.go:1378]: You can do this `DeleteFile` as an operation on the `pfc` which will make the entire commit atomic and not require calling `StartCommit` and `FinishCommit`.

File: src/client/pps/pps.proto (hunk 1 above)
Comment (ysimonson) [reply src/client/pps/pps.proto:42]: The builders section of the PR description gives some outline, and you can see the official builders in `etc/pipeline-build`. In short, the Pachyderm client isn't included at the moment (though maybe it should be?) Rather, it's just the language's official docker images, with some glue shell scripts that are expected to execute in the build and run phases.

File: src/server/pps/cmds/cmds.go (hunk 2 above)
Comment (ysimonson) [reply src/server/pps/cmds/cmds.go:1378]: Neat!

**Pull request merged**